### ISSUE TITLE
parse_config extension question

### ISSUE THREAD

**Issue by frederic-loui** [napalm-automation/napalm-yang#118]
Hi !
I'm still studying openconfig combined with ansible and napalm. I fully support the conclusion of this [article](http://blog.ipspace.net/2017/05/start-using-openconfig-with-napalm-on.html) as to use napalm as a transition mechanism (waiting for the vendor to support the OC models completely ). I tried various example and wrote based on the network-instance model the following example of a router with ISIS enabled:
```
result I would like to achieve when getting config from a Junos device
{
    "network-instances": {
        "network-instance": {
            "global": {
                "name": "global", 
                "protocols": {
                    "protocol": {
                        "isis 2200": {
                            "identifier": "isis", 
                            "isis": {
                                "interfaces": {
                                    "interface": {
                                        "ge-0/0/3.0": {
                                            "interface-id": "ge-0/0/3.0", 
                                            "config": {
                                                "circuit-type": "POINT_TO_POINT"
                                            }
                                        }, 
                                        "ge-0/0/2.0": {
                                            "interface-id": "ge-0/0/2.0", 
                                            "config": {
                                                "circuit-type": "POINT_TO_POINT"
                                            }
                                        }, 
                                        "ge-0/0/1.0": {
                                            "interface-id": "ge-0/0/1.0", 
                                            "config": {
                                                "circuit-type": "POINT_TO_POINT"
                                            }
                                        }
                                    }
                                }, 
                                "global": {
                                    "config": {
                                        "net": [
                                            "49.4225.0000.0000.5555.00"
                                        ]
                                    }
                                }, 
                                "levels": {
                                    "level": {
                                        "1": {
                                            "level-number": "1", 
                                            "config": {
                                                "metric-style": "WIDE_METRIC", 
                                                "enabled": true
                                            }
                                        }
                                    }
                               }
                            }, 
                            "name": "2200"
                        }
                    }
                }
            }
        }
    }
```

I used napalm-yang to retrieve ISIS config with model "openconfig_network_instance" from devices:
```
...
model.add_model(napalm_yang.models.openconfig_network_instance)
...
```

And got the following result with `(model.translate_config(profile=['junos'])`: 

```
{
    "network_instances": {
        "network-instance": {
            "global": {
                "name": "global", 
                "config": {
                    "type": "DEFAULT_INSTANCE", 
                    "enabled": True
                }, 
                "protocols": {
                    "protocol": {
                        "isis isis": {
                            "name": "isis", 
                            "identifier": "isis"
                        } 
                    }
                }
            } 
        }
    }
}
```

while the config on the the Junos vmx is:
```
mgmt@gn4-net-core-5> show configuration protocols isis 
level 1 wide-metrics-only;
level 2 wide-metrics-only;
interface ge-0/0/0.0 {
    point-to-point;
}
interface ge-0/0/1.0 {
    point-to-point;
}
interface ge-0/0/2.0 {
    point-to-point;
}
interface ge-0/0/3.0 {
    point-to-point;
}
interface lo0.0;

interfaces {
    lo0 {
            family iso {
                address 49.4225.0000.0000.5555.00;
            }
        }
    }
}

protocols {
    isis {
        level 1 wide-metrics-only;
         interface ge-0/0/0.0 {
            point-to-point;
        }
        interface ge-0/0/1.0 {
            point-to-point;
        }
        interface ge-0/0/2.0 {
            point-to-point;
        }
        interface ge-0/0/3.0 {
            point-to-point;
        }
        interface lo0.0;
    }
}

```

Clearly `parse_config` needs some extensions to be implemented. So the question is:
Is there a documented/recommended process that I can follow so that Junos -> OC and OC -> Junos for `parse_config` can take into account the "full OC model" ?

(Sorry if it is a stupid/silly question that has been previously answered, in that case I'd love to have some pointers :-) )

Thanks for your great job !
All the best,

Frederic

**Comment by tcaiazza** [comment]
This is because the parser mapping IS-IS in Junos hasn't been written yet.  While napalm-yang maybe able to generate the oc model that you would like, it doesn't necessarily mean that the model is being parsed correctly.  Even for the parts that are said to be supported here https://napalm-yang.readthedocs.io/en/latest/root/supported_models/index.html not everything works, yet.

Now if you want to start working on getting this parsed and contributing it back, here is the mapping file https://github.com/napalm-automation/napalm-yang/blob/develop/napalm_yang/mappings/junos/parsers/config/openconfig-network-instance/includes/isis.yaml

I would also read through the docs here, https://napalm-yang.readthedocs.io/en/latest/developers/parsers/index.html to understand a bit more how it works.   

Also looking through other completed parser files helps, like the oc-interface model.  https://github.com/napalm-automation/napalm-yang/blob/develop/napalm_yang/mappings/junos/parsers/config/openconfig-interfaces/interfaces.yaml

Let me know if you get stuck and I can try to help, I've started to write some parsers for eos and nxos.

**Comment by dbarrosop** [comment]
Nothing to add. Thanks @tcaiazza :D

Just one thing; @frederic-loui, the learning curve is a bit steep at the beginning because of the SDL but you get a lot of things for free thanks to it. I'd advise you to work on a branch of your own with some mocked data (you can see some tests here already for an example on how to set everything for easy testing: https://github.com/napalm-automation/napalm-yang/tree/develop/test/integration/test_profiles/junos/openconfig-network-instance/config/default). That way if you get stuck you can push what you have and let us know and we will try to get you unblocked.

**Comment by frederic-loui** [comment]
Hi all,
Thanks for your help and pointer. I managed to populate the ISIS model via the isis parser initiated by [dbarrosop](https://github.com/dbarrosop). And manage to get the following result from my test router lab: 
```
{
    "network_instances": {
        "network-instance": {
            "global": {
                "name": "global", 
                "config": {
                    "type": "DEFAULT_INSTANCE", 
                    "enabled": True
                }, 
                "protocols": {
                    "protocol": {
                         "isis isis": {
                            "isis": {
                                "interfaces": {
                                    "interface": {
                                        "ge-0/0/0.0": {
                                            "interface-id": "ge-0/0/0.0"
                                        }, 
                                        "ge-0/0/1.0": {
                                            "interface-id": "ge-0/0/1.0"
                                        }, 
                                        "ge-0/0/2.0": {
                                            "interface-id": "ge-0/0/2.0"
                                        }, 
                                        "ge-0/0/3.0": {
                                            "interface-id": "ge-0/0/3.0"
                                        }, 
                                        "lo0.0": {
                                            "interface-id": "lo0.0"
                                        }
                                    }
                                }
                            }, 
                            "name": "isis", 
                            "identifier": "isis"
                        } 
                    }
                }
            }
        }
    }
}
```

I just won the first battle ... not the war (yet :-) )
I will work on completing as much as possible the model and provide you a pull request.

If you want just to check & confirm that I'm on the right track just please check [here](https://github.com/frederic-loui/napalm-yang/blob/develop/napalm_yang/mappings/junos/parsers/config/openconfig-network-instance/includes/isis.yaml): 
(I did not wrote a lot but enough to get a satisfying result as a parser Noob :-) )

In any case, thanks for your support !

All the best,
Frederic

**Comment by frederic-loui** [comment]
Hi again, 
While parsing ISIS JUNOS config I encountered the following problem related to `ISIS metric style`. In Junos using wide metric is done via the following config stanza (in XML):

```
    <configuration junos:commit-seconds="1513852112" junos:commit-localtime="2017-12-21 10:28:32 UTC" junos:commit-user="mgmt">
            <protocols>
                <isis>
                    <level>
                        <name>1</name>
                        <wide-metrics-only/>
                    </level>
                    <level>
                        <name>2</name>
                        <wide-metrics-only/>
                    </level>

... output omitted for clarity

              </isis>
            </protocols>
    </configuration>
    <cli>
        <banner></banner>
    </cli>
</rpc-reply>
```
So my problem here is to find the syntax related to solve the following problem:

If if find `<wide-metrics-only/>` => the value of the `oc-netinst:metric-style` is `WIDE_METRIC` else it is `NARROW_METRIC1.

I saw some example using the map keyword but in my case i need to test the presence of an XML stanza (a well know string "`wide-metrics-only`") but if i find no XML "`wide-metrics-only`" then I set `oc-netinst:metric-style` to `NARROW_METRIC`

**Comment by dbarrosop** [comment]
Well, your solution actually looks good so let me test and see what could be happening there.

**Comment by dbarrosop** [comment]
Would you mind pushing to your branch everything you have already done and also provide some mocked data as I did in PR #120? Just populate the file `cli.1._get_configuration_.0`, please, the rest can be empty for now.

**Comment by frederic-loui** [comment]
For now, I just modified isis.yaml file you can find the changes [here](https://github.com/frederic-loui/napalm-yang/blob/develop/napalm_yang/mappings/junos/parsers/config/openconfig-network-instance/includes/isis.yaml)

I worked directly on a vSRX device in my LAB so that I can tweak config on the fly.
Here is the config:
```
<rpc-reply xmlns:junos="http://xml.juniper.net/junos/15.1X49/junos">
    <configuration junos:commit-seconds="1516705583" junos:commit-localtime="2018-01-23 11:06:23 UTC" junos:commit-user="mgmt">
            <version>15.1X49-D50.3</version>
            <system>
                <host-name>gn4-net-core-5</host-name>
                <root-authentication>
                </root-authentication>
                <login>
                    <class>
                        <name>CLASS-OPERATOR-READ-ONLY</name>
                        <permissions>access</permissions>
                        <permissions>view</permissions>
                        <permissions>view-configuration</permissions>
                        <allow-commands>show</allow-commands>
                    </class>
                    </login>
                    <services>
                        <ssh>
                        </ssh>
                        <netconf>
                            <ssh>
                            </ssh>
                        </netconf>
                        <web-management>
                            <http>
                                <interface>fxp0.0</interface>
                            </http>
                        </web-management>
                    </services>
                    <syslog>
                        <user>
                            <name>*</name>
                            <contents>
                                <name>any</name>
                                <emergency/>
                            </contents>
                        </user>
                        <file>
                            <name>messages</name>
                            <contents>
                                <name>any</name>
                                <any/>
                            </contents>
                            <contents>
                                <name>authorization</name>
                                <info/>
                            </contents>
                        </file>
                        <file>
                            <name>interactive-commands</name>
                            <contents>
                                <name>interactive-commands</name>
                                <any/>
                            </contents>
                        </file>
                    </syslog>
                    <license>
                        <autoupdate>
                            <url>
                                <name>https://ae1.juniper.net/junos/key_retrieval</name>
                            </url>
                        </autoupdate>   
                    </license>
                </system>
                <security>
                    <forwarding-options>
                        <family>
                            <mpls>
                                <mode>packet-based</mode>
                            </mpls>
                        </family>
                    </forwarding-options>
                    <zones>
                        <security-zone>
                            <name>trust</name>
                            <interfaces>
                                <name>ge-0/0/0.0</name>
                                <host-inbound-traffic>
                                    <system-services>
                                        <name>snmp</name>
                                    </system-services>
                                </host-inbound-traffic>
                            </interfaces>
                        </security-zone>
                    </zones>
                </security>
                <interfaces>
                    <interface>
                        <name>ge-0/0/0</name>
                        <unit>
                            <name>0</name>
                            <description>R5@ge-0/0/0.0-&gt;R1-ge-0/0/2.0</description>
                            <family>
                                <inet>
                                    <address>
                                        <name>172.16.15.5/24</name>
                                    </address>
                                </inet>
                                <iso>
                                </iso>
                                <mpls>
                                </mpls>
                            </family>
                        </unit>
                    </interface>
                    <interface>
                        <name>ge-0/0/1</name>
                        <unit>
                            <name>0</name>
                            <disable/>
                            <description>R5@ge-0/0/1.0-&gt;R2@ge-0/0/2.0</description>
                            <family>
                                <inet>
                                    <address>
                                        <name>172.16.25.5/24</name>
                                    </address>
                                </inet>
                                <iso>
                                </iso>
                                <mpls>
                                </mpls>
                            </family>
                        </unit>
                    </interface>
                    <interface>
                        <name>ge-0/0/2</name>
                        <unit>
                            <name>0</name>
                            <description>R5@ge-0/0/2.0-&gt;R3@ge-0/0/2.0</description>
                            <family>
                                <inet>
                                    <address>
                                        <name>172.16.35.5/24</name>
                                    </address>
                                </inet>
                                <iso>
                                </iso>
                                <mpls>
                                </mpls>
                            </family>
                        </unit>
                    </interface>
                    <interface>
                        <name>ge-0/0/3</name>
                        <unit>
                            <name>0</name>
                            <description>R5@ge-0/0/3.0-&gt;R4@ge-0/0/2.0</description>
                            <family>
                                <inet>
                                    <address>
                                        <name>172.16.45.5/24</name>
                                    </address>
                                </inet>
                                <iso>
                                </iso>
                                <mpls>
                                </mpls>
                            </family>
                        </unit>
                    </interface>
                    <interface>
                        <name>fxp0</name>
                        <unit>
                            <name>0</name>
                            <family>
                                <inet>
                                    <address>
                                        <name>10.134.250.35/9</name>
                                    </address>
                                </inet>
                            </family>
                        </unit>
                    </interface>
                    <interface>
                        <name>lo0</name>
                        <unit>
                            <name>0</name>
                            <family>
                                <inet>
                                    <address>
                                        <name>172.16.5.5/32</name>
                                    </address>
                                </inet>
                                <iso>
                                    <address>
                                        <name>49.4225.0000.0000.5555.00</name>
                                    </address>
                                </iso>
                            </family>
                        </unit>
                        <unit>
                            <name>4225</name>
                            <family>
                                <inet>
                                    <address>
                                        <name>5.5.5.5/32</name>
                                    </address>
                                </inet>
                            </family>
                        </unit>
                    </interface>
                </interfaces>
                <snmp>
                    <location>gn4-net-core-5</location>
                    <contact>llopat@man.poznan.pl</contact>
                    <community>
                        <name>public</name>
                        <authorization>read-only</authorization>
                        <clients>
                            <name>10.134.250.2/32</name>
                        </clients>
                        <clients>
                            <name>0.0.0.0/0</name>
                            <restrict/>
                        </clients>
                    </community>
                </snmp>
                <routing-options>
                    <router-id>172.16.5.5</router-id>
                    <autonomous-system>
                        <as-number>65525</as-number>
                    </autonomous-system>
                </routing-options>
                <protocols>
                    <mpls>
                        <interface>
                            <name>ge-0/0/0.0</name>
                        </interface>
                        <interface>
                            <name>ge-0/0/1.0</name>
                        </interface>
                        <interface>
                            <name>ge-0/0/2.0</name>
                        </interface>
                        <interface>
                            <name>ge-0/0/3.0</name>
                        </interface>
                    </mpls>
                    <bgp>
                        <group>
                            <name>iBGP-&gt;RR-CLIENT</name>
                            <type>internal</type>
                            <description>iBGP-&gt;RR-CLIENT/GN4-NET-CORE-4</description>
                            <local-address>172.16.5.5</local-address>
                            <family>
                                <inet>
                                    <unicast>
                                    </unicast>
                                </inet>
                                <inet-vpn>
                                    <unicast>
                                    </unicast>
                                </inet-vpn>
                            </family>
                            <cluster>172.16.5.5</cluster>
                            <neighbor>
                                <name>172.16.1.1</name>
                            </neighbor>
                            <neighbor>
                                <name>172.16.2.2</name>
                            </neighbor>
                            <neighbor>
                                <name>172.16.3.3</name>
                            </neighbor>
                            <neighbor>
                                <name>172.16.4.4</name>
                            </neighbor>
                        </group>
                    </bgp>
                    <isis>
                        <level>
                            <name>1</name>
                            <wide-metrics-only/>
                        </level>        
                        <level>
                            <name>2</name>
                            <preference>50</preference>
                        </level>
                        <interface>
                            <name>ge-0/0/0.0</name>
                            <point-to-point/>
                        </interface>
                        <interface>
                            <name>ge-0/0/1.0</name>
                            <point-to-point/>
                        </interface>
                        <interface>
                            <name>ge-0/0/2.0</name>
                            <point-to-point/>
                        </interface>
                        <interface>
                            <name>ge-0/0/3.0</name>
                            <point-to-point/>
                        </interface>
                        <interface>
                            <name>lo0.0</name>
                        </interface>
                    </isis>
                    <ldp>
                        <transport-address>
                            <router-id/>
                        </transport-address>
                        <interface>
                            <name>ge-0/0/0.0</name>
                        </interface>
                        <interface>
                            <name>ge-0/0/1.0</name>
                        </interface>
                        <interface>
                            <name>ge-0/0/2.0</name>
                        </interface>
                        <interface>
                            <name>ge-0/0/3.0</name>
                        </interface>
                        <neighbor>
                            <name>172.16.1.1</name>
                        </neighbor>
                        <neighbor>
                            <name>172.16.2.2</name>
                        </neighbor>
                        <neighbor>
                            <name>172.16.3.3</name>
                        </neighbor>
                        <neighbor>
                            <name>172.16.4.4</name>
                        </neighbor>
                    </ldp>              
                </protocols>
                <policy-options>
                    <policy-statement>
                        <name>NMAAS-ADMIN-AS65525-IMPORT</name>
                        <term>
                            <name>BGP</name>
                            <from>
                                <protocol>bgp</protocol>
                                <community>NMAAS-ADMIN-AS65525-COMMUNITY</community>
                            </from>
                            <then>
                                <accept/>
                            </then>
                        </term>
                        <term>
                            <name>CONNECTED</name>
                            <from>
                                <protocol>direct</protocol>
                                <community>NMAAS-ADMIN-AS65525-COMMUNITY</community>
                            </from>
                            <then>
                                <accept/>
                            </then>
                        </term>
                    </policy-statement>
                    <community>
                        <name>NMAAS-ADMIN-AS65525-COMMUNITY</name>
                        <members>target:65525L:65525</members>
                    </community>
                </policy-options>
                <routing-instances>
                    <instance>
                        <name>NMAAS-ADMIN-AS65525</name>
                        <instance-type>vrf</instance-type>
                        <interface>
                            <name>lo0.4225</name>
                        </interface>
                        <route-distinguisher>
                            <rd-type>172.16.5.5:65525</rd-type>
                        </route-distinguisher>
                        <vrf-import>NMAAS-ADMIN-AS65525-IMPORT</vrf-import>
                        <vrf-target>
                            <community>target:65525L:65525</community>
                        </vrf-target>
                        <vrf-table-label>
                        </vrf-table-label>
                    </instance>
                </routing-instances>
        </configuration>
        <cli>
            <banner>[edit]</banner>
        </cli>
    </rpc-reply>                        
```

**Comment by frederic-loui** [comment]
Hi, sorry I misread your previous message. I put the configuration sample in [cli.1._get_configuration_.0](https://github.com/frederic-loui/napalm-yang/blob/develop/test/integration/test_profiles/junos/openconfig-network-instance/config/default/mocked/cli.1._get_configuration_.0)

**Comment by dbarrosop** [comment]
No worries. For now it's ok but before sending a PR you can create a new test case by creating a new folder:

`napalm-yang/test/integration/test_profiles/junos/openconfig-network-instance/config/is-is/`

and replicating the structure in `default` (where you put the configuration file)

**Comment by dbarrosop** [comment]
ok, so I have fixed your issue here:

https://github.com/napalm-automation/napalm-yang/pull/121/files#diff-ae7d2163057d243402219fa126909b7d

Note the only difference with what you intuitively figured it was the solution is the `present: true` keyword. Basically you don't care about the value of the leaf (which is none), you care about the presence of it.

**Comment by frederic-loui** [comment]
Great I'll sync my repo in order to get your updates.
Thanks !

Le 23 janv. 2018 4:51 PM, "David Barroso" <notifications@github.com> a
écrit :

**Comment by dbarrosop** [comment]
ok, I think I finally got this right, see: https://github.com/napalm-automation/napalm-yang/pull/122

Note that I removed the `value` directive because it was conflating some pre/post actions and replaced it with two different directives; pre/post.

**Comment by frederic-loui** [comment]
Great! Will test it and get back to you ASAP. Thanks !

**Comment by dbarrosop** [comment]
I am closing this one, feel free to reopen if you still need help for this.
